cl/lib/pacer.py

```diff
@@ -530,3 +534,118 @@ def normalize_attorney_contact(c, fallback_name=""):
+
+    :param court_id: The court ID to check.
+    :returns: A dict with the court connection status.
+    """
+
+    url = f"https://ecf.{court_id}.uscourts.gov/"
+
+    connection_ok = False
+    status_code = None
+    try:
+        r = requests.get(url, timeout=5)
+        if r.status_code == requests.codes.ok:
+            connection_ok = True
+        status_code = r.status_code
+    except requests.exceptions.RequestException as e:
+        connection_ok = False
+
+    blocked_dict: ConnectionType = {
+        "connection_ok": connection_ok,
+        "status_code": status_code,
+        "date_time": datetime.now(timezone.utc),
+    }
+    return blocked_dict
+
+
+def get_or_cache_court_status(court_id: str, server_ip: str) -> bool:
+    """Get the court status from Redis or cache it if it's not there.
+
+    :param court_id: The court ID to check.
+    :param server_ip: The server IP address.
+    :return: True if connection was successful, False otherwise.
+    """
+
+    court_status_key = "status:pacer:court.%s:ip.%s"
+    r = make_redis_interface("CACHE", decode_responses=False)
+    pickle_status = r.get(court_status_key % (court_id, server_ip))
+    if pickle_status:
+        court_status = pickle.loads(pickle_status)
+        return court_status
+
+    # Unable to find court_status in cache, getting it from request.
+    connection_info = check_pacer_court_connectivity(court_id)
+    current_status = connection_info["connection_ok"]
+
+    # Stores court connection status with court ID and server IP as key.
+    # 10 minutes expiration time.
+    status_expiration = 60 * 10
+    court_key = court_status_key % (court_id, server_ip)
+    r.set(court_key, pickle.dumps(current_status), ex=status_expiration)
+    if connection_info["connection_ok"]:
+        return True
+
+    # If court connection failed, log the error and return False.
+    log_the_problem(connection_info, court_id, server_ip)
+    return False
+
+
+def log_the_problem(
+    connection_info: ConnectionType,
+    court_id: str,
+    server_ip: str,
+) -> None:
+    """Log the problem with the court in Redis.
+
+    :param connection_info: A dict as returned by
+    check_pacer_court_connectivity method
+    :param court_id: The court ID
+    :param server_ip: The server IP address
+    :return: None
+    """
+    r = make_redis_interface("STATS")
+    pipe = r.pipeline()
+    d = connection_info["date_time"].date().isoformat()
+    t = connection_info["date_time"].time().isoformat()
+    ip_key = f"pacer_log:c:{court_id}:server_ip:{server_ip}:d:{d}.ip_error"
+
+    status_code = connection_info["status_code"]
+    if status_code is None:
+        status_code = 0
+    if connection_info["connection_ok"] is True:
+        connection_ok = "True"
+    else:
+        connection_ok = "False"
+
+    pipe.hset(ip_key, "connection_ok", connection_ok)
+    pipe.hset(ip_key, "status_code", status_code)
+    pipe.hset(ip_key, "time", t)
```

Comment:
I think you could use the `mapping` parameter of HSET to do this in one command, no:

https://redis-py.readthedocs.io/en/stable/commands.html#redis.commands.core.CoreCommands.hset

So, maybe just update the `connection_info` dict, and then pass it?

Reply:
I tried to use mapping to set the values in one line and it worked, however with this change mypy failed, because I'm using a TypedDict for the connection_info dict so seems this is not compatible with mapping: 

![Screen Shot 2022-06-16 at 13 50 16](https://user-images.githubusercontent.com/486004/174151392-13bfe367-1249-4354-b4dc-33bf416948b6.png)

Reply:
This is fixed, seems that `HSET` mapping parameter expects a Mapping[str | bytes, str | int] type, so I had to previously convert the dict to this type of object and it passed mypy verification.

Reply:
Hm, I wonder if you could use `cast` here and get what we need, but I'm going to merge anyhow. I just don't like changing code this much to make the type checker happy.